### DISCUSSION TITLE
Driver Enable(DE) and Receiver Enable(RE) control in ModbusRTU

### DISCUSSION THREAD

**Discussion by Devendra-7** [edgexfoundry/community#238] [Q&A]
Hi All,
 I have enabled the device-modbus(-go) service in imx6 board and able to start the service with necessary configuration for the modbus RTU communication, but the device-modbus service is not able to read the data from my slave device.
It's not throwing any error related to the configuration and slave device accessibility.
want to know whether DE& RE pin control is included in the device service ? or we need to implement same thing in device service?.

P.S:  Communication between master(imx) and modbus slave device is verified by controlling the DE & RE pins in C application(without edgex device-services). libmodbus library is used for developing the Modbus RTU C -application.

Thanks in Advance....!

**Comment by cloudxxx8** [comment]
Could you provide more reference about DE&RE? 
In theory, if it's a device path in the Linux system, it should be able to connect.
However, the open source device-modbus is a reference implementation, so we won't support additional feature enhancement.
You can write your own Modbus Device Service based on Go or C SDK, or you may consider to use the comercial version.

**Reply by Devendra-7** [reply]
Hi @cloudxxx8,

 Driver Enable(DE) and Receiver Enable(RE) in a MAX485 chip is to control the transmission and receiving of data from a modbus slave device.
 It's not a device path. We are using IIoT Gateway board, which is having MAX485 chip with RE& DE pins. I have ported the Edgex device services i.e device-modbus in Gateway board which act like modbus master, to read the data from slave device we need to control the DE&RE pins accordingly, to get the data from slave device.
 We have verified the reading process by using custom C application, by controlling the DE& RE pins with help of GPIO's. 
 
 Current device-modbus  service implementation doesn't have the support to control the DE&RE pins, device-modbus service working fine with the master's which is having Auto direction control feature in MAX485 chip sets(in this case no need to control the DE&RE pins)

**Reply by eaton-coreymutter** [reply]
It's difficult in Linux to get good control of these from userspace, especially at higher baud rates, since you don't know quite how long it takes between your write() and when the bytes actually finish getting sent out the line.

Lately my team has been using devices that have auto-RTS support in the RS485 UART, and that works well. (RTS controlling DE and not(RTS) controlling RE)

The one Modbus gateway my team made that didn't have auto-RTS, we ended up making a custom tty driver for, so we would know when the transmit was complete (the "tx empty" interrupt), and control RTS from the driver. That worked OK, but needed a bunch of tweaking to get the timings just right.
